### ISSUE TITLE
Example with multiple drones

### ISSUE THREAD

**Issue by julianoes** [mavlink/MAVSDK-Python#102]
It would be good to provide an example how we can use the Python wrappers to connect to multiple drones at the same time.

It's a bit tricky because we need to run a `mavsdk_server` for each drone.

**Comment by JonasVautherin** [comment]
> It's a bit tricky because we need to run a mavsdk_server for each drone.

Or maybe we should extend the API to support multiple drones like in MAVSDK-C++. We talked about that a while ago, if I remember correctly :blush:.

**Comment by HanbumKo** [comment]
Is there any update?? Or is there any reference issue of this??

**Comment by julianoes** [comment]
No there is currently no update from our side as it is not our priority.

Would you be able to contribute/work on this?

**Comment by HanbumKo** [comment]
Yeah I could. But I need to understand this system and codes before doing this. Is there any additional development documentations of MAVSDK? Thanks.

**Comment by julianoes** [comment]
You have probably seen this? https://mavsdk.mavlink.io/develop/en/cpp/

Do you already understand the C++ API, gRPC backend and language frontends? If not I'd suggest to look at the C++ API first and see how multiple drones are working there.

**Comment by HanbumKo** [comment]
I'm not familiar with gRPC. OK let me see C++ API first and would do that. But now I've been focusing on other project, so I couldn't say 'I can!!'. Thanks!!

**Comment by traverseda** [comment]
> > It's a bit tricky because we need to run a mavsdk_server for each drone.
> 
> Or maybe we should extend the API to support multiple drones like in MAVSDK-C++. We talked about that a while ago, if I remember correctly blush.

So it looks to me like currently if you have several mavSDK systems on one ip/port (for example you're using qgroundcontrol's MAVlink forwarding) it picks a random device to communicate with. This is pretty surprising behavior and could easily be a safety issue. I suspect it gets weirder than that, but I didn't continue testing past that point.

Even just keeping the current system where you need multiple mavsdk_server's but allowing you to specify the system-id of the drone you're trying to communicate with would be a big improvement.

**Comment by JonasVautherin** [comment]
> if you have several mavSDK systems on one ip/port

Assuming that the different systems have a different IP, you could write a simple demultiplexer that would redirect each drone (based on its ip) to a different mavsdk_server instance :+1:.

**Comment by traverseda** [comment]
I believe that mavlink-router can accomplish this, it's still another jenga block on the tower though and makes it a fair bit harder to develop nice self-container apps that are reliable.

**Comment by JonasVautherin** [comment]
I don't think mavlink-router does this :thinking:, though it does a lot of other things that you don't need here.

However, such a multiplexer is quite trivial to write (I wrote something similar in ~80 lines of python), and it can totally be part of your main python app, right?

**Comment by traverseda** [comment]
How? Are you using pymavlink for that?

**Comment by JonasVautherin** [comment]
No, I just proxy the UDP packets based on the source IP (I assume that each drone has a different IP, and redirect the messages of each drone to the corresponding `mavsdk_server` instance).

**Comment by traverseda** [comment]
Oh, I misunderstood. That's not going to work with qgroundcontrol's mavlink forwarding. I'd prefer to keep the drones connected to both qgroundcontrol and my script for safety reasons.

**Comment by JonasVautherin** [comment]
> I don't really understand what the proxy would be doing here that can't be done by just having mavsdk directly connect to each of those IPs separately.

Well the proxy allows MAVSDK to connect to each IP separately, by having one MAVSDK instance per IP :thinking:. It seems like you would like to have multiple drones sending to the same UDP port on MAVSDK, but IMO that is a bad idea (and that's not how mavsdk_server works).

> That's not going to work with qgroundcontrol's mavlink forwarding.

Right, I don't know your setup, so I couldn't say.

**Comment by traverseda** [comment]
> but IMO that is a bad idea (and that's not how mavsdk_server works).

It is however how the actual CPP mavsdk API appears to wrok, where nodes get identified by their sysid's and you can subscribe to an event listener that tells you when a new drone shows up. The functionality exists, it's just exposed over the grpc interface.

Just saying, it's been about 4 years since this issue was opened, and there are still very much use-cases where it's a problem. I can go in to more detail about the specifics of those use cases if needed. The 1000 foot overview is that the default PX4 simulator script uses a UDP-broadcast mode that doesn't seem to work with a direct connection, trying to use a forwarder like the one built-in to qgroundcontrol won't work, there's no real way to auto-detect multiple drones under python, etc. It introduces a fair bit of friction and makes the python SDK unsuitable for making a real app.

This is a bit of an upstream problem as it exists in the mavsdk_server binary implementation and can't really be fixed on the python side of things.

**Comment by JonasVautherin** [comment]
> It is however how the actual CPP mavsdk API appears to wrok

It is, and MAVLink allows it. I am just saying that IMO it is a bad idea over a UDP/IP network :see_no_evil:. It is easy to handle in C++, but harder to translate to the MAVSDK gRPC system. So I am reluctant to spend a lot of my own time solving this particular problem in this particular way, because I personally believe that it is generally not the right way.

Now if you find a way to add support for that over gRPC in a clean way, then I will be more than happy to review your idea/PR :+1:.

**Comment by traverseda** [comment]
If I had the ability to specify what sysid a system is trying to connect to, and a method that let me know when new sysid's became available, I could hack something together using the existing spawn-a-new-server-for-each-drone methodology. That *should* bring feature parity to the python SDK for auto discovery. I can probably even submit some upstream PR's to the python side of things to wrap it in a nicer interface.

From the python side of things I'd imagine something like

```python

drones={}
startport=50051

watcher=System(port=startport)

async for newdrone in watcher.find_new_broadcasters("udp://:14550"):
    startport+=1
    drone=System(port=startport)
    await drone.connect(system_address=newdrone["system_address"],sysid=newdrone["sysid"])
   drones[newdrone["sysid"]]=drone

```

Not familiar enough with the internals to know if this makes sense, but would something like that work without breaking your grpc model? Doesn't require you to pass complex objects around or state the sysid in every message.

**Comment by JonasVautherin** [comment]
Unfortunately, `mavsdk_server` only really starts when the drone is detected: https://github.com/mavlink/MAVSDK/blob/main/src/mavsdk_server/src/mavsdk_server_api.cpp#L13. So it's not straightforward to add detection logic there.

To be honest, if I had time to improve that, I would rather spend it working on the direct FFI wrappers (i.e. removing gRPC and calling the C++ API directly from the language bindings). This would bring the C++ discovery to the wrappers. But that is not trivial to do.

**Comment by drtrigon** [comment]
We do work with the python (gRPC) library and also encounter the limitations. Thus I am looking for an alternative - as I understand the way to go would be C++. We would prefer to use CSharp instead, but it is declared as "Proof of concept." on the mavsdk webpage. Can you tell me whether it uses gRPC (like python) and suffers from the same limitations or is it like C++ and has full access to all features?

**Comment by JonasVautherin** [comment]
C# is not maintained at all, I would advise against using it.

This said, an easy way to support multiple drones is to run multiple instances of MAVSDK and have the drones send MAVLink on different ports. I would (and I did in the past, actually) write a simple proxy that would receive messages from the drones on 14550 and that would redirect them to a new port.

You could maybe ask e.g. @ThomasDebrunner if Auterion would be willing to open source the reverse proxy I wrote, I believe it does exactly that.

**Comment by drtrigon** [comment]
It might be verry possible that we got that reverse proxy already 😉 (however we modified it a bit meanwhile)

We do run multiple servers and it works fine. However occasionally we run into other limitations as e.g. waypoints do not support autocontinue param in python where they do in C++, etc. Also scalability might become an issue at some point. Currently we would like to operate at least 3-4 drones. Thus we are looking for alternatives. Would it make sense to use ctypes from python?


Am 10. April 2024 12:55:54 MESZ schrieb Jonas Vautherin ***@***.***>:

**Comment by Brennanzuz** [comment]
> > if you have several mavSDK systems on one ip/port
> 
> Assuming that the different systems have a different IP, you could write a simple demultiplexer that would redirect each drone (based on its ip) to a different mavsdk_server instance 👍.

I'm facing the same issue trying to get MAVSDK-Python to get to work with ArduPilot's `sim_vehicle.py` SITL with multiple vehicles which uses MAVProxy to squish all of the drones' data into one port (eg. `--out=localhost:14540`) which MAVSDK doesn't take very well. 

So I made a demultiplexer in Python:

```python
import asyncio

MIXED_TRAFFIC_PORT: int = 14540
MAVSDK_SERVER_ADDRESS: str = "127.0.0.1"
MAVSDK_SERVER_PORTS: dict[int, int] = {
    1: 14551,
    2: 14552,
    3: 14553
}

class DemuxProtocol(asyncio.DatagramProtocol):
    """
    Receives a mix of packets bound for different sys_ids from the same address
    and splits them up to be sent to the respective MAVSDK server ports.
    """
    def __init__(self):
        # Maps sys_id to the corresponding MAVSDK server socket. This is the
        # same object as the bound sockets in ReverseProtocol.transport
        self.transports: dict[int, asyncio.DatagramTransport] = {}
        # Maps sys_id to the actual (ephemeral) ports of each drone hidden
        # behind the address
        self.drone_addresses: dict[int, tuple] = {}

    def connection_made(self, transport):
        self.transport = transport

    def datagram_received(self, data, addr):
        # Extract the sys_id from the packet
        if len(data) < 6:
            return
        # MAVLink v1 format: [0xFE, len, seq, sys_id, comp_id, msg_id, ...]
        if data[0] == 0xFE:
            sys_id = data[3]
        # MAVLink v2 format: [0xFD, len, incompat, compat, seq, sys_id, ...]
        elif data[0] == 0xFD:
            sys_id = data[5]
        else:
            return

        # Learns the actual ephemeral address of each connected drone at runtime
        self.drone_addresses[sys_id] = addr

        # Send to the corresponding MAVSDK server
        if sys_id in self.transports:
            self.transports[sys_id].sendto(data, None)

class ReverseProtocol(asyncio.DatagramProtocol):
    """Receives MAVSDK command packets and forwards them back to the drone."""
    def __init__(self, demux_proto: DemuxProtocol, sys_id: int):
        # Reference back to the unbound main socket to send back to the drone
        self.demux_proto: DemuxProtocol = demux_proto
        self.sys_id: int = sys_id

    def connection_made(self, transport):
        self.transport = transport

    def datagram_received(self, data, addr):
        # We have received a packet from this drone's MAVSDK server (for
        # example, a script sending a request for telemetry)
        drone_address = self.demux_proto.drone_addresses.get(self.sys_id)
        if drone_address:
            # Use the unbound main socket (that has reference to the actual
            # drones' ports) to specify the right address to send back
            self.demux_proto.transport.sendto(data, drone_address)

async def run_demux():
    loop = asyncio.get_running_loop()
    demux_protocol = DemuxProtocol()
    # Set up all sockets
    await loop.create_datagram_endpoint(
        lambda: demux_protocol,
        local_addr=("0.0.0.0", MIXED_TRAFFIC_PORT)
    )
    for sys_id, port in MAVSDK_SERVER_PORTS.items():
        rev = ReverseProtocol(demux_protocol, sys_id)
        transport, _ = await loop.create_datagram_endpoint(
            lambda r=rev: r,
            remote_addr=(MAVSDK_SERVER_ADDRESS, port)
        )
        demux_protocol.transports[sys_id] = transport
    await asyncio.sleep(float("inf"))

if __name__ == "__main__":
    asyncio.run(run_demux(), debug=True)
```

Running this, messages (eg. telemetry) from ArduPilot's drones can be distinguished based on their sys_id despite being on the same address and port, and commands from MAVSDK bound for the drones will be appropriately sent back to the drones on the SITL.

Hope this helps someone!
